patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,758 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
+Date: Mon, 1 Aug 2022 22:50:29 -0400
+Subject: [PATCH] Brigadier based command API
+
+
+diff --git a/build.gradle.kts b/build.gradle.kts
+index 279a666e8ea2c07f41ee3f28b768e95dca5f0a10..51ca5d873e109e435255cc8a1204764b00e7adf3 100644
+--- a/build.gradle.kts
++++ b/build.gradle.kts
+@@ -46,7 +46,7 @@ dependencies {
+     implementation("org.ow2.asm:asm:9.4")
+     implementation("org.ow2.asm:asm-commons:9.4")
+     // Paper end
+-
++    api("com.mojang:brigadier:1.0.18") // Paper, expose!
+     api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
+     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
+     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..46bf23740aaa23ae51b1f5dbd8cf3c467afa7750
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
+@@ -0,0 +1,94 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.CommandDispatcher;
++import com.mojang.brigadier.builder.ArgumentBuilder;
++import com.mojang.brigadier.tree.CommandNode;
++import net.kyori.adventure.util.Services;
++import org.bukkit.Bukkit;
++import org.bukkit.plugin.Plugin;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++@ApiStatus.NonExtendable
++public abstract class CommandBuilder extends ArgumentBuilder<CommandSourceStack, CommandBuilder> {
++
++    /**
++     * Creates a new command builder.
++     *
++     * @param plugin plugin
++     * @param name   command name
++     * @return builder
++     */
++    @NotNull
++    public static CommandBuilder of(@NotNull Plugin plugin, @NotNull String name) {
```

Comment:
I think people are moving away from the `of` methods because they don't work well with static imports. It should be smth like `commandBuilder` instead.

Reply:
I think that `CommandBuilder.commandBuilder` is the more common pattern. Adventure uses it frequently, e.g. `ClickEvent.clickEvent`.